File: src/ai_guardrails/generators/lefthook.py
"""LefthookGenerator — tamper-protected lefthook.yml generation.

Generates a lefthook.yml with pre-commit hooks appropriate for the
detected languages. Always includes security hooks (gitleaks,
suppress-comments, validate-configs). Adds language-specific hooks
when the corresponding language is in the detected set.
"""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Any, ClassVar

import yaml

from ai_guardrails.generators.base import (
    HASH_HEADER_PREFIX,
    compute_hash,
    verify_hash,
)

if TYPE_CHECKING:
    from ai_guardrails.models.registry import ExceptionRegistry

# ---------------------------------------------------------------------------
# Hook definitions
# ---------------------------------------------------------------------------

_ALWAYS_HOOKS: dict[str, Any] = {
    "suppress-comments": {
        "glob": "*.{py,js,ts,tsx,jsx,rs,cs,go,lua,sh}",
        "run": (
            "uv run python -m ai_guardrails.hooks.suppress_comments {staged_files}"
        ),
        "priority": 2,
    },
    "protect-configs": {
        "glob": (
            "ruff.toml|biome.json|lefthook.yml|.editorconfig|.markdownlint.jsonc|.codespellrc"
        ),
        "run": ("uv run python -m ai_guardrails.hooks.protect_configs {staged_files}"),
        "priority": 2,
    },
    "gitleaks": {
        "run": "gitleaks protect --staged --no-banner",
        "priority": 2,
    },
    "detect-secrets": {
        "glob": "!.secrets.baseline",
        "run": "detect-secrets-hook --baseline .secrets.baseline {staged_files}",
        "priority": 2,
    },
    "codespell": {
        "glob": "*.{py,md,txt,yaml,yml,toml,json}",
        "run": "codespell --check-filenames {staged_files}",
        "priority": 2,
    },
    "markdownlint": {
        "glob": "*.md",
        "run": "markdownlint-cli2 {staged_files}",
        "priority": 2,
    },
    "validate-configs": {
        "run": "uv run python -m ai_guardrails generate --check",
        "priority": 2,
    },
}

_PYTHON_HOOKS: dict[str, Any] = {
    "python-format-and-stage": {
        "glob": "*.py",
        "run": (
            "uv run ruff format {staged_files}"
            " && uv run ruff check --fix {staged_files}"
            " && git add {staged_files}"
        ),
        "stage_fixed": True,
        "priority": 1,
    },
    "ruff-check": {
        "glob": "*.py",
        "run": "uv run ruff check {staged_files}",
        "priority": 2,
    },
}

_SHELL_HOOKS: dict[str, Any] = {
    "shell-format-and-stage": {
        "glob": "*.{sh,bash,zsh}",
        "run": "shfmt -i 2 -ci -bn -w {staged_files} && git add {staged_files}",
        "stage_fixed": True,
        "priority": 1,
    },
    "shellcheck": {
        "glob": "*.{sh,bash,zsh}",
        "run": "shellcheck --severity=info -x {staged_files}",
        "priority": 2,
    },
}

_COMMIT_MSG_HOOKS: dict[str, Any] = {
    "conventional": {
        "run": (
            r"grep -qE "
            r'"^(feat|fix|docs|style|refactor|perf|test|build|ci|chore|revert)'
            r"(\\(.+\\))?!?:"
            r'" {1}'
        ),
    },
}


def _build_commands(languages: list[str]) -> dict[str, Any]:
    """Assemble pre-commit commands based on detected languages."""
    commands: dict[str, Any] = dict(_ALWAYS_HOOKS)
    if "python" in languages:
        commands.update(_PYTHON_HOOKS)
    if "shell" in languages:
        commands.update(_SHELL_HOOKS)
    return commands


class LefthookGenerator:
    """Generates tamper-protected lefthook.yml with language-appropriate hooks."""

    name = "lefthook"
    output_files: ClassVar[list[str]] = ["lefthook.yml"]

    def generate(
        self,
        registry: ExceptionRegistry,  # noqa: ARG002
        languages: list[str],
        project_dir: Path,  # noqa: ARG002
    ) -> dict[Path, str]:
        """Return {Path("lefthook.yml"): content} with hash header."""
        body = self._build_body(languages)
        header = f"{HASH_HEADER_PREFIX}{compute_hash(body)}"
        content = header + "\n" + body
        return {Path("lefthook.yml"): content}

    def check(
        self,
        registry: ExceptionRegistry,  # noqa: ARG002
        project_dir: Path,
    ) -> list[str]:
        """Return stale/missing descriptions (empty list = fresh)."""
        target = project_dir / "lefthook.yml"
        if not target.exists():
            return ["lefthook.yml is missing — run: ai-guardrails generate"]
        existing = target.read_text(encoding="utf-8")
        # check() has no language list — detect staleness against no-language base
        body = self._build_body([])
        if not verify_hash(existing, body):
            return ["lefthook.yml is stale or tampered — run: ai-guardrails generate"]
        return []

    def _build_body(self, languages: list[str]) -> str:
        """Build the lefthook.yml body from language set."""
        config: dict[str, Any] = {
            "pre-commit": {"commands": _build_commands(languages)},
            "commit-msg": {"commands": _COMMIT_MSG_HOOKS},
        }
        return yaml.dump(config, default_flow_style=False, sort_keys=False)

File: src/generators/lefthook.ts
import type { ResolvedConfig } from "@/config/schema";
import type { ConfigGenerator } from "@/generators/types";
import type { LanguagePlugin } from "@/languages/types";
import { withHashHeader } from "@/utils/hash";

/**
 * Convert a glob pattern to a regex string suitable for lefthook's `exclude` field.
 * Lefthook excludes accept a Go regex, not a glob.
 *
 * Conversion rules:
 * - `**` → `.*`
 * - `*` → `[^/]*`
 * - `.` → `\.`
 * - Other regex metacharacters are escaped
 */
function globToRegex(glob: string): string {
  // Escape regex metacharacters except `*` (handled separately)
  const escaped = glob.replace(/[.+^${}()|[\]\\?]/g, (ch) =>
    ch === "?" ? "[^/]" : `\\${ch}`
  );
  // Replace `**` before `*` to avoid double-processing
  return escaped.replace(/\*\*/g, ".*").replace(/\*/g, "[^/]*");
}

function buildExcludeRegex(ignorePaths: readonly string[]): string | undefined {
  if (ignorePaths.length === 0) return undefined;
  const parts = ignorePaths.map(globToRegex);
  return parts.length === 1 ? (parts[0] ?? "") : `(${parts.join("|")})`;
}

function renderLefthookYml(
  activePluginIds: ReadonlySet<string>,
  ignorePaths: readonly string[]
): string {
  const hasPython = activePluginIds.has("python");
  const hasTs = activePluginIds.has("typescript");
  const excludeRegex = buildExcludeRegex(ignorePaths);
  const excludeLine =
    excludeRegex !== undefined ? `\n      exclude: '${excludeRegex}'` : "";

  const pythonSection = hasPython
    ? `
    ruff-fix:
      glob: "*.py"${excludeLine}
      run: ruff check --fix {staged_files} && git add {staged_files}
      stage_fixed: true
      priority: 1
`
    : "";

  const tsSection = hasTs
    ? `
    biome-fix:
      glob: "*.{ts,tsx,js,jsx}"${excludeLine}
      run: biome check --write {staged_files} && git add {staged_files}
      stage_fixed: true
      priority: 1
`
    : "";

  return `pre-commit:
  commands:
${pythonSection}${tsSection}
    gitleaks:
      run: gitleaks protect --staged --no-banner
      fail_text: "Potential secrets detected"
      priority: 2

    codespell:
      glob: "*.{ts,md,txt,yaml,yml,toml,json,py,go,rs}"
      exclude: 'tests/fixtures/.*${excludeRegex !== undefined ? `|${excludeRegex}` : ""}'
      run: codespell --check-filenames {staged_files}
      fail_text: "Spell errors found"
      priority: 2

    markdownlint:
      glob: "*.md"${excludeLine}
      run: markdownlint-cli2 {staged_files}
      fail_text: "Markdown lint errors found"
      priority: 2

    check-suppress-comments:
      glob: "*.{py,ts,tsx,js,jsx,rs,go,cs,lua,sh,bash,zsh,ksh,c,cpp,cc,h,hpp}"${excludeLine}
      run: ai-guardrails hook suppress-comments {staged_files}
      fail_text: "Inline suppression comments require a reason"
      priority: 2

    no-commits-to-main:
      run: |
        branch=$(git rev-parse --abbrev-ref HEAD 2>/dev/null) || exit 0
        if [ "$branch" = "main" ] || [ "$branch" = "master" ]; then
          echo "Direct commits to main are not allowed"
          exit 1
        fi
      fail_text: "Do not commit directly to main"
      priority: 2

commit-msg:
  commands:
    conventional:
      run: grep -qE "^(feat|fix|docs|style|refactor|perf|test|build|ci|chore|revert)(\\(.+\\))?!?:" {1}
      fail_text: "Commit message must follow Conventional Commits format"
`;
}

/**
 * Generate a lefthook.yml tailored to the active language plugins.
 * Callers must pass active plugin ids resolved from detectLanguages().
 */
export function generateLefthookConfig(
  config: ResolvedConfig,
  activePlugins: readonly LanguagePlugin[]
): string {
  const ids = new Set(activePlugins.map((p) => p.id));
  return withHashHeader(renderLefthookYml(ids, config.ignorePaths));
}

export const LEFTHOOK_GENERATOR_ID = "lefthook";

export const lefthookGenerator: ConfigGenerator = {
  id: LEFTHOOK_GENERATOR_ID,
  configFile: "lefthook.yml",
  generate(_config: ResolvedConfig): string {
    throw new Error(
      "lefthookGenerator.generate() must not be called directly — use generateLefthookConfig(config, languages) instead"
    );
  },
};
